drift/lib/src/runtime/manager/references.dart

```diff
@@ -0,0 +1,491 @@
+///
+/// # Prefetching
+///
+/// ### Problem
+///
+/// It's quite common that a user will want a model together with it's relations.
+/// The simple option is to prefetch all of the users, and then run a 2nd request for each related group.
+/// ```dart
+/// final users = await users.withReferences().get()
+/// for (final (user,refs) in users){
+///   final group = await refs.groups.getSingle();
+/// }
+/// ```
+/// The issue with the above code is that we could be setting ourselves up to do tons of queries.
+///
+/// ### Solution
+///
+/// There are two methods we use together to solve this.
+/// Joins and Prefetches.
+///
+/// ##### Join
+///
+/// For fields which are foreign keys to other tables, we can just use a join to select the related field and get it all in one query
+/// There are many examples how to do this in the drift documentation
+///
+/// ##### Prefetch
+///
+/// However, for reverse relations, we can't get it all in a single query.
+/// We have to run 2 queries. Here is how it's done manually.
+///
+/// ```dart
+/// final groups = await groups.get();
+/// final groupIds = groups.map((group)=> group.id);
+/// final users = await users.filter((f)=> f.group.id.isIn(groupIds)).get()
+/// final groupsWithUsers = groups.map((group)=>(group,users.where((user)=> user.group == group.id)));
+/// ```
+///
+/// We get all the users ahead of time, and then manually return each group with it's users using a `where(...)` filter.
+///
+/// We will call both of these "prefetch" throughout.
+///
+///
+/// ### Manager API
+///
+/// This is quite verbose, and the manager api seeks to solve this for you.
+/// The API looks like this:
+/// ```dart
+/// users.withReferences((prefetch) => prefetch(groups: true)).get()
+/// ```
+///
+/// To do this there are a couple of things we need to do.
+///
+/// For references data we are getting via JOIN
+///
+/// 1. Before we even run a query with `users.withReferences((prefetch) => prefetch(group: true)).get()`, we need to put add a JOIN to this query.
+/// 2. When we read back this result, we need to read the JOINed information back.
+///
+/// For referenced data we are getting with a prefetch:
+///
+/// 1. After we get all the results we need to run more queries, map the correct data, and inject it info each object.
+///
+/// We also have to do difference behavior depending of what was prefetched, so this is a pretty complex issue.
+///
+/// ### Implementation
+///
+/// We only need to do 2 things for this to work:
+///
+/// 1) Reference class
+///
+/// This class is kinda cool. It holds the original result of the query as a `TypedResult`, which is what drift uses to store the raw query.
+/// But along with that, it has getters for prebuilt managers.
+/// When `withReferences` is called, we return the dataclass, together with this.
+///
+/// To keep things simple for now, this is a reference class which has it's prefetching abilities removed
+/// ```dart
+/// final class $$ProductTableReferences
```

Comment:
I think it would be helpful to show a consistent example throughout this explainer - so probably the users / groups relation?

Reply:
Or updating the initial examples to use the products table might be easier than updating the larger examples at the end :D